compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertDmaOutOfOrderBlock.cpp

```diff
@@ -0,0 +1,202 @@
+    return connectionOp.emitOpError()
+           << "expected exactly one source channel for connection op";
+  }
+  auto channelOp = dyn_cast<AMDAIE::ChannelOp>(srcChannels[0].getDefiningOp());
+  if (!channelOp)
+    return connectionOp.emitOpError() << "expected an `amdaie.channel` op";
+  if (mm2sChannelToDmaStarts.count(channelOp) == 0) {
+    return connectionOp.emitOpError()
+           << "source channel does not have corresponding MM2S DMA start op";
+  }
+  if (mm2sChannelToDmaStarts[channelOp].size() != 1) {
+    return connectionOp.emitOpError()
+           << "expected exactly one MM2S DMA start op";
+  }
+  return mm2sChannelToDmaStarts[channelOp][0];
+}
+
+LogicalResult insertDmaOutOfOrder(
+    IRRewriter &rewriter, AMDAIE::ChannelOp s2mmChannelOp,
+    SmallVector<AMDAIE::DMAStartOp> &s2mmDmaStartOps,
+    DenseMap<AMDAIE::ChannelOp, SmallVector<AMDAIE::DMAStartOp>>
+        &mm2sChannelToDmaStarts) {
+  // If there is only one DMA start op, do nothing.
+  if (s2mmDmaStartOps.size() <= 1) return success();
+  // Merge the given DMA start ops into a single out-of-order block.
+  int8_t newRepeatCount = 0;
+  SmallVector<Value> newConnections;
+  SmallVector<SmallVector<Operation *>> lockAndBdOps;
+  for (AMDAIE::DMAStartOp s2mmDmaStartOp : s2mmDmaStartOps) {
+    // Reject the merge if the DMA start op already has more than one connection
+    // or if it is not packet-switched.
+    SmallVector<Value> connections = s2mmDmaStartOp.getConnections();
+    newConnections.append(connections.begin(), connections.end());
+    if (connections.size() != 1) {
+      return s2mmDmaStartOp.emitOpError()
+             << "expected exactly one connection for DMA start op, but found "
+             << connections.size() << "\n";
+    }
+    AMDAIE::ConnectionOp connectionOp =
+        dyn_cast<AMDAIE::ConnectionOp>(connections[0].getDefiningOp());
+    if (connectionOp.getConnectionType() != AMDAIE::ConnectionType::Packet) {
+      return s2mmDmaStartOp.emitOpError()
+             << "out-of-order mode requires DMA to be packet-switched\n";
+    }
+    // Find the corresponding MM2S DMA start op.
+    // TODO(zhewen): Add support for npu.write_bd when the MM2S channel is
+    // located in the shim tile.
+    FailureOr<AMDAIE::DMAStartOp> maybeMm2sDmaStartOp =
+        getMm2sDmaStart(connectionOp, mm2sChannelToDmaStarts);
+    if (failed(maybeMm2sDmaStartOp)) {
+      return s2mmDmaStartOp.emitOpError()
+             << "could not find corresponding MM2S DMA start op";
+    }
+    // Get the blocks in the S2MM and MM2S DMA start ops.
+    SmallVector<Block *> s2mmBlocks = llvm::to_vector(llvm::map_range(
+        s2mmDmaStartOp.getBody().getBlocks(), [](Block &b) { return &b; }));
+    SmallVector<Block *> mm2sBlocks = llvm::to_vector(
+        llvm::map_range(maybeMm2sDmaStartOp.value().getBody().getBlocks(),
+                        [](Block &b) { return &b; }));
+    int8_t numBDs = 0;
+    for (size_t i = 0; i < s2mmBlocks.size(); i++) {
```

Comment:
I think you could just iterate over the blocks?
```suggestion
    for (Block *block : s2mmBlocks) {
```
And could you add a comment explaining this block of logic?

Reply:
Need to iterate on both the S2MM and MM2S sides, so changed to `for (auto &&[s2mmBlock, mm2sBlock] : llvm::zip(s2mmBlocks, mm2sBlocks))` instead